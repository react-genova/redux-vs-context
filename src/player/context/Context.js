import { createContext } from 'react';
import { createSelector } from 'reselect';

const PlayerContext = createContext({
    name: '',
    surname: '',
    onNameChanged: () => null,
    onSurnameChanged: () => null
});

const getPlayer = value => value;
export const getPlayerName = createSelector(
    getPlayer,
    player => player.name
);
export const getPlayerSurname = createSelector(
    getPlayer,
    player => player.surname
);
export const getPlayerNameChangedCallback = createSelector(
    getPlayer,
    player => player.onNameChanged
);
export const getPlayerSurnameChangedCallback = createSelector(
    getPlayer,
    player => player.onSurnameChanged
);

export default PlayerContext;
