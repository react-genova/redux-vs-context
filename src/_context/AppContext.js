import React from 'react';
import Context from '../player/context/Context';
import { useState, useCallback } from 'react';

const AppContext = ({ children }) => {
    const [name, setName] = useState('');
    const [surname, setSurname] = useState('');
    const onNameChanged = useCallback(name => setName(name));
    const onSurnameChanged = useCallback(surname => setSurname(surname));
    return (
        <Context.Provider
            value={{ name, surname, onNameChanged, onSurnameChanged }}
        >
            {children}
        </Context.Provider>
    );
};

export default AppContext;
