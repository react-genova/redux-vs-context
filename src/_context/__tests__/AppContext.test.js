import React, { Fragment } from 'react';
import { render, fireEvent } from 'react-testing-library';
import AppContext from '../AppContext';
import Context from '../../player/context/Context';

const toEvent = value => ({ target: { value } });

describe('AppContext provider component', () => {
    it('provides an empty value at startup and reacts to changes', () => {
        const { container, getByTestId } = render(
            <AppContext>
                <Context.Consumer>
                    {({ name, surname, onNameChanged, onSurnameChanged }) => (
                        <Fragment>
                            <input
                                data-testid="name"
                                onChange={({ target: { value } }) =>
                                    onNameChanged(value)
                                }
                                value={name}
                            />
                            <input
                                data-testid="surname"
                                onChange={({ target: { value } }) =>
                                    onSurnameChanged(value)
                                }
                                value={surname}
                            />
                        </Fragment>
                    )}
                </Context.Consumer>
            </AppContext>
        );
        // check snapshot
        expect(container).toMatchSnapshot();
        // retrieves rendered components
        const name = getByTestId('name');
        const surname = getByTestId('surname');
        // expect values to be empty at startup
        expect(name.value).toBe('');
        expect(surname.value).toBe('');
        // simulate a change on name
        fireEvent.change(name, toEvent('Paolino'));
        // expect name to be changed and surname not
        expect(name.value).toBe('Paolino');
        expect(surname.value).toBe('');
        // simulate a change on surname
        fireEvent.change(surname, toEvent('Paperino'));
        // expect surname to be changed and name not
        expect(name.value).toBe('Paolino');
        expect(surname.value).toBe('Paperino');
    });
});
