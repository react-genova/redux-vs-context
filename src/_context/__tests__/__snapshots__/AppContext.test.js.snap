// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContext provider component provides an empty value at startup and reacts to changes 1`] = `
<div>
  <input
    data-testid="name"
    value=""
  />
  <input
    data-testid="surname"
    value=""
  />
</div>
`;
